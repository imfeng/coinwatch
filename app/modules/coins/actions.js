// @flow

import { CALL_HANDLER } from '../../middleware/handler';
import { getCoinListApi, getCoinPriceApi } from '../../connectors/cryptocompare/api';

export const GET_COINS_ATTEMPT = 'GET_COINS_ATTEMPT';
export const GET_COINS_SUCCESS = 'GET_COINS_SUCCESS';
export const GET_COINS_FAILURE = 'GET_COINS_FAILURE';

export function getCoins() {
  return {
    [CALL_HANDLER]: {
      types: [GET_COINS_ATTEMPT, GET_COINS_SUCCESS, GET_COINS_FAILURE],
      handler: getCoinListApi,
    },
  };
}

export const GET_COIN_PRICE_ATTEMPT = 'GET_COIN_PRICE_ATTEMPT';
export const GET_COIN_PRICE_SUCCESS = 'GET_COIN_PRICE_SUCCESS';
export const GET_COIN_PRICE_FAILURE = 'GET_COIN_PRICE_FAILURE';

export function getCoinPrice(coin) {
  return {
    [CALL_HANDLER]: {
      types: [GET_COIN_PRICE_ATTEMPT, GET_COIN_PRICE_SUCCESS, GET_COIN_PRICE_FAILURE],
      handler: getCoinPriceApi,
      data: coin.Symbol,
    },
  };
}
