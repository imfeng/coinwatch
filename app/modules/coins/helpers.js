export function generateSlug(from, to, exchange) {
  return `${from}-${to}-${exchange}`;
}

export default { generateSlug };
