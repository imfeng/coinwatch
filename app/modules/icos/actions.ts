// @flow

import { CALL_HANDLER } from '../../middleware/handler';
import { getIcosListApi } from '../../connectors/icowatchlist/api';

export const GET_ICOS_ATTEMPT = 'GET_ICOS_ATTEMPT';
export const GET_ICOS_SUCCESS = 'GET_ICOS_SUCCESS';
export const GET_ICOS_FAILURE = 'GET_ICOS_FAILURE';

export function getIcos() {
  return {
    [CALL_HANDLER]: {
      types: [GET_ICOS_ATTEMPT, GET_ICOS_SUCCESS, GET_ICOS_FAILURE],
      handler: getIcosListApi,
    },
  };
}
