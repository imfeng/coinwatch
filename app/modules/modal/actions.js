export const OPEN_MODAL = 'OPEN_MODAL';

export function openModal(component) {
  return (dispatch) => {
    dispatch({ type: OPEN_MODAL, component });
  };
}

export const CLOSE_MODAL = 'CLOSE_MODAL';

export function closeModal() {
  return (dispatch) => {
    dispatch({ type: CLOSE_MODAL });
  };
}
