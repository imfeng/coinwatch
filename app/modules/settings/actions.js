import settings from 'electron-settings';
import { findIndex, find, forEach } from 'lodash';

export const SETTINGS_SAVE_COIN_ATTEMPT = 'SETTINGS_SAVE_COIN_ATTEMPT';
export const SETTINGS_SAVE_COIN_SUCCESS = 'SETTINGS_SAVE_COIN_SUCCESS';
export const SETTINGS_SAVE_COIN_FAILED = 'SETTINGS_SAVE_COIN_FAILED';
export function saveCoinSettings(coin) {
  return (dispatch) => {
    dispatch({ type: SETTINGS_SAVE_COIN_ATTEMPT });

    if (!settings.has('coins')) {
      settings.set('coins', []);
    }
    const coins = settings.get('coins');

    if (find(coins, (c) => c.coin.Id === coin.coin.Id)) {
      return dispatch({ type: SETTINGS_SAVE_COIN_FAILED });
    }

    coins.push(coin);
    settings.set('coins', coins);

    return dispatch({ type: SETTINGS_SAVE_COIN_SUCCESS });
  }
}

export const SETTINGS_REMOVE_COIN_ATTEMPT = 'SETTINGS_REMOVE_COIN_ATTEMPT';
export const SETTINGS_REMOVE_COIN_SUCCESS = 'SETTINGS_REMOVE_COIN_SUCCESS';
export const SETTINGS_REMOVE_COIN_FAILED = 'SETTINGS_REMOVE_COIN_FAILED';
export function removeCoinSettings({ Id }) {
  return (dispatch) => {
    dispatch({ type: SETTINGS_REMOVE_COIN_ATTEMPT });

    const coins = settings.get('coins');
    const index = findIndex(coins, (c) => c.coin.Id === Id);
    if (index === -1) {
      return dispatch({ type: SETTINGS_REMOVE_COIN_FAILED });
    }

    coins.splice(index, 1);

    settings.set('coins', coins);

    return dispatch({ type: SETTINGS_REMOVE_COIN_SUCCESS });
  };
}

export const SETTINGS_MAIN_UPDATE = 'SETTINGS_MAIN_UPDATE';
export function updateMainSettings(data) {
  forEach(data, (value, key) => {
    settings.set(key, value);
  });

  return (dispatch) => {
    dispatch({ type: SETTINGS_MAIN_UPDATE });
  };
}

export const SETTINGS_AUTOLAUNCH_ON = 'SETTINGS_AUTOLAUNCH_ON';
export const SETTINGS_AUTOLAUNCH_OFF = 'SETTINGS_AUTOLAUNCH_OFF';
export function updateAutolaunchSettings(data) {
  if (data === "yes") {
    return (dispatch) => {
      dispatch({ type: SETTINGS_AUTOLAUNCH_ON });
    };
  }

  return (dispatch) => {
    dispatch({ type: SETTINGS_AUTOLAUNCH_OFF });
  };
}
