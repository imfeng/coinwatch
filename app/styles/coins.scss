.coins {
  padding: 10px;

  h3 {
    text-transform: uppercase;
    text-align: center;
    color: #000;
  }

  thead > tr {
    background-color: #333;

    a {
      color: #fff;

      &:hover {
        color: #ccc;
      }
    }
  }

  .coin {
    background-color: $table-tr-background-color;
    color: #000;
  }

  .coin:nth-of-type(odd) {
    background-color: $table-tr-background-color-odd;
  }

  .coin {
    //display: flex;

    & > div {
      flex-grow: 1;
    }

    .name {
      //display: flex;
      //flex-direction: row;
      //text-align: left;
      //flex-grow: 2;

      img {
        width: 30px;
        height: 30px;
      }

      span {
        margin-left: 5px;
        margin-top: 5px;
      }
    }

    .price {
      //display: flex;
      //flex-direction: row;
      margin: 5px 50px;
    }

    .actions {
      margin-top: 5px;
      color: red;
      text-align: right;

      & > a {
        margin: 0 5px;
      }

      .remove {
        &:hover {
          color: red;
        }
      }
    }
  }
}

form {
  input[type="submit"] {
    background: $navbar-color;
    color: #fff;
    margin: 10px auto;
    width: 100%;

    &:hover {
      background: lighten($navbar-color, 15);
      color: #fff;
    }
  }
}
