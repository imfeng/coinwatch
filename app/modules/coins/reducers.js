// @flow

import settings from 'electron-settings';
import { cloneDeep, values, map, findIndex, find } from 'lodash';
import { ipcRenderer } from 'electron';
import {
  GET_COINS_SUCCESS,
  GET_COIN_PRICE_SUCCESS,
  GET_EXCHANGE_LIST_SUCCESS,
  GET_SYMBOL_LIST_SUCCESS,
  SETTINGS_REMOVE_COIN_SUCCESS,
  SETTINGS_SAVE_COIN_SUCCESS,
  SETTINGS_TOGGLE_VISIBILITY_SUCCESS,
} from './actions';
import formatPrice from '../../helpers/formatPrice';
import { TRAY_UPDATE } from '../../main/Tray';

const initialState = {
  data: [],
  exchanges: [],
  symbols: [],
  coins: settings.get('coins') || [],
};

export default function coinsReducer(state = initialState, action) {
  let coins = null;
  let coin = null;
  switch (action.type) {
    case SETTINGS_TOGGLE_VISIBILITY_SUCCESS:
      coins = cloneDeep(state.coins);
      coin = find(coins, (c) => c.slug === action.data.slug);

      coin.visibility = !coin.visibility;

      settings.set('coins', coins);

      return { ...state, coins };
      break;
    case SETTINGS_SAVE_COIN_SUCCESS:
      return { ...state, coins: settings.get('coins') };
      break;
    case SETTINGS_REMOVE_COIN_SUCCESS:
      return { ...state, coins: settings.get('coins') };
      break;
    case GET_COINS_SUCCESS:
      return { ...state, data: action.response };
      break;
    case GET_EXCHANGE_LIST_SUCCESS:
      return { ...state, exchanges: action.response };
      break;
    case GET_SYMBOL_LIST_SUCCESS:
      return { ...state, symbols: action.response };
      break;
    case GET_COIN_PRICE_SUCCESS:
      coins = cloneDeep(state.coins);
      coin = find(coins, (c) => c.slug === action.data.slug);

      if (!coin) {
        return { ...state };
      }

      coin.price = map(action.response, formatPrice).join('');

      ipcRenderer.send(TRAY_UPDATE, coins);

      return { ...state, coins };
      break;
    default:
      return state;
  }
}
