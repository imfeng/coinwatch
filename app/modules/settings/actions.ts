import settings from 'electron-settings';
import { Dispatch } from 'redux';
import { JsonObject } from '../../types/JsonValue';

export const SETTINGS_MAIN_UPDATE = 'SETTINGS_MAIN_UPDATE';

export async function updateMainSettings(data: JsonObject) {
  for (const key in data) {
    await settings.set(key, data[key]);
  }

  return (dispatch: Dispatch) => {
    dispatch({ type: SETTINGS_MAIN_UPDATE });
  };
}

export const SETTINGS_AUTOLAUNCH_ON = 'SETTINGS_AUTOLAUNCH_ON';
export const SETTINGS_AUTOLAUNCH_OFF = 'SETTINGS_AUTOLAUNCH_OFF';

export function updateAutolaunchSettings(data: string) {
  if (data === 'yes') {
    return (dispatch: Dispatch) => {
      dispatch({ type: SETTINGS_AUTOLAUNCH_ON });
    };
  }

  return (dispatch: Dispatch) => {
    dispatch({ type: SETTINGS_AUTOLAUNCH_OFF });
  };
}
