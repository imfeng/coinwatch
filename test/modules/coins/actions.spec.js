import { spy } from 'sinon';
import * as actions from '../../../app/modules/coins/actions';

describe('actions', () => {
  it('should get the coins list', () => {
    expect(actions.getCoins()).toMatchSnapshot();
  });

  it('should get the coin price from a symbol', () => {
    const coin = {
      Symbol: 'ETH',
    };
    expect(actions.getCoinPrice(coin)).toMatchSnapshot();
  });
});
