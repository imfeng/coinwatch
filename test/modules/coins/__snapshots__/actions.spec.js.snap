// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should get the coin price from a symbol 1`] = `
Object {
  Symbol(Call Handler): Object {
    "data": "ETH",
    "handler": [Function],
    "types": Array [
      "GET_COIN_PRICE_ATTEMPT",
      "GET_COIN_PRICE_SUCCESS",
      "GET_COIN_PRICE_FAILURE",
    ],
  },
}
`;

exports[`actions should get the coins list 1`] = `
Object {
  Symbol(Call Handler): Object {
    "handler": [Function],
    "types": Array [
      "GET_COINS_ATTEMPT",
      "GET_COINS_SUCCESS",
      "GET_COINS_FAILURE",
    ],
  },
}
`;
